docs/propagation.md

```diff
@@ -210,3 +217,39 @@ along F0, propagate `["c"]` along F1, and propagate nothing along F2.
 T0  | "a", **"b"** | **"c"**  | "f" |
 T1  | "a", "b"     | "c", "d" | "g" |
 T2  | **"a", "b"** | "c", "e" |     |
+
+### Data flow edge ops
+
+The above propagation step description applies to most ops. However, there are
+cases where a sharding rule is not appropriate. For those cases, Shardy defines
+_data flow edge_ ops.
+
+A data flow edge of some op X defines a bridge between a set of *sources* and a
+set of *targets*, such that all sources and targets should be sharded in the
+same way. Examples of such ops are `OptimizationBarrierOp`, `WhileOp`, and
+`CaseOp`. Ultimately, any op that implements [ShardableDataFlowOpInterface](sdy_op_interfaces#shardabledataflowopinterface_shardabledataflowopinterface)
+is considered a data flow edge op.
+
+An op can have multiple data flow edges that are orthogonal to one another. For
+example:
+
+```mlir
+    y_0, ..., y_n = while (x_0, ..., x_n)
+                    ((pred_arg_0,... , pred_arg_n) { ... })
+                    ((body_arg_0,..., body_arg_n) {
+                    ...
+                    return return_value_0, ..., return_value_n
+                    })
+```
+
+This while op has `n` data flow edges: the i-th data flow edges is between
+sources `x_i`, `return_value_i` and targets `y_i`, `pred_arg_i`,
+`body_arg_i`.
+
+Propagation will propagate shardings between all sources and targets of a
+`sdy.data_flow_edge` as if it was a regular op with the sources as operands
+and targets as results, and an identity `sdy.op_sharding_rule`. That means
+that forward propagation is from sources to targets and backwards
+propagation is from targets to sources.
+
+See [sdy.data_flow_edge](./sdy_dialect.md#sdy.data_flow_edge) for more information.
```

Comment:
As indicated above, this op is an impl detail, so lets not link to it for more information. You already linked to ShardableDataFlowOpInterface, which should be enough :)